### PR TITLE
ssh agent: add timeout, flags, improve error handling.

### PR DESCRIPTION
**Pull request by ccontavalli**
This PR adds timeouts in communicationg with the ssh-agent and improves
on the error handling and reporting. On top of it, it plumbs in flags to
better control ssh-agent behavior (allowing for example, to pass extra
flags to the ssh agent, as well as to point to a wrapper, useful to start
setting -t, -c, or to allow for a wrapper), or just to simplify testing.

Commits are fairly independent and self contained, the series includes
some generic improvements: go fmt, updating testify, adding array flags, ...

including a fix for `enkit login` not starting the agent properly.

- lib/kauth: solve enkit login bug.
- lib/kflags: add support for array of strings passed as arguments.
- lib/kflags: run go fmt in the directory.
- proxy/ptunnel/commands/agent: drop deprecated string, noop.
- lib/kcerts: improve error messages and comments.
- lib/kcerts: go fmt
- lib/kcerts: add configurable timeout in handling the ssh agent.
- lib/kcerts: separate object state from serialized state.
- dependencies: update testify library to the latest version.
- lib/kcerts: add more ssh agent flags, plumb flags throughout, add tests.

### CODE REVIEW

**Comment by ccontavalli** [comment]
The failure seems unrelated to this PR.

**Comment by ccontavalli** [comment]
@minor-fixes btw, I've been using this for the last several days, no issue so far.

**Review by jonathan-enf — APPROVED** [review]

**Comment by jonathan-enf** [comment]
> The failure seems unrelated to this PR.

Maybe the issue is the merge conflicts?

File: lib/kcerts/ssh.go [hunk 1]
```diff
@@ -166,11 +217,13 @@ type AgentCert struct {
 
 // Principals returns a map where the keys are the CA's PKS and the certs identities are the values
 func (a SSHAgent) Principals() ([]AgentCert, error) {
-	conn, err := net.Dial("unix", a.Socket)
+	conn, err := net.DialTimeout("unix", a.Socket, a.Timeout)
```
Comment (minor-fixes) [inline lib/kcerts/ssh.go]: Should this have a context plumbed through, use `DialContext` instead?

**Review by minor-fixes — APPROVED** [review]

File: lib/kcerts/ssh.go [hunk 2]
```diff
@@ -82,18 +82,26 @@ func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) er
...
+// state about an existing ssh-agent.
+type SSHAgentState struct {
 	PID    int    `json:"pid"`
 	Socket string `json:"sock"`
+}
+
+// SSHAgent is a wrapper around golang.org/x/crypto/ssh/agent to ease the
+// creation and management of ssh-agents.
+type SSHAgent struct {
+	SSHAgentState
```
[the remark this answers is not in the document]
Comment (ccontavalli) [reply lib/kcerts/ssh.go:96]: as, for example, `state SSHAgentState`? no objection, yes, was trying to minimized the diff.

File: lib/kcerts/ssh.go (hunk 1 above)
Comment (ccontavalli) [reply lib/kcerts/ssh.go]: "plumbed through" as passed by the caller? I think there's value in the caller not having to hard code a timeout or have its own flags / mechanics to configure a context, and just assume the library it is invoking enforces some sane behavior (can invoke the agent directly if something fancier is needed).

If you mean to just change a.Timeout to a context with deadline... I have no strong feelings around it. Either way it works. Let me know.

**Pull request merged**